### PR TITLE
LRS2 lipreading recipe

### PR DESCRIPTION
**Pull request by LiChenda**
Add a lipreading recipe for  LRS2 dataset.

The recipe use a pretrained model by myself to extract visual features as the asr input feature.
The pretrained model is a 3-D resnet frontend in a Isolated-word lipreading model. 
The training procedure follows the paper _"Stafylakis, T., & Tzimiropoulos, G. (2017). Combining residual networks with LSTMs for lipreading."_
The extracted features are saved in kaldi feats scp and ark style in `data.sh`.
Currently, the default config files have not been carefully finetuned. 
I may upload a pretrained lip reading model later, and maybe people can tune better models with this recipe.

### CODE REVIEW

**Comment by sw005320** [comment]
@ftshijt, could I ask you to review this PR?
This is not a standard ASR recipe, but many parts are still common.

File: egs2/lrs2/lipreading1/local/data.sh [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    done
+fi
+
+if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
+    # Download the pretrained model to extract visual features.
+    # The model was trained by Chenda Li (lichenda1996@sjtu.edu.cn), 
+    # following the paper by Stafylakis, T., & Tzimiropoulos, G. (2017). 
+    # "Combining residual networks with LSTMs for lipreading".
+    if [ ! -f ./local/feature_extract/finetuneGRU_19.pt ]; then
+        wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=12ww9Vp8q3g-PdNGKvgEW8dPmFlJdhko0' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=12ww9Vp8q3g-PdNGKvgEW8dPmFlJdhko0" -O ./local/feature_extract/finetuneGRU_19.pt.tgz && rm -rf /tmp/cookies.txt
```
Comment (sw005320) [inline egs2/lrs2/lipreading1/local/data.sh:48]: as I mentioned, it would be better to upload it through zenodo.

**Review by sw005320 — COMMENTED** [review]
Thanks, @LiChenda!

Can you tell me where local/feature_extract/concatenate_feature.py is used?

File: egs2/lrs2/lipreading1/local/data.sh (hunk 1 above)
Comment (LiChenda) [reply egs2/lrs2/lipreading1/local/data.sh]: Done.

**Comment by LiChenda** [comment]
> Thanks, @LiChenda!
> 
> Can you tell me where local/feature_extract/concatenate_feature.py is used?

Sorry, it has been deprecated, I forgot to remove it. Now it is removed from this PR.

File: egs2/lrs2/lipreading1/db.sh [hunk 2]
```diff
@@ -0,0 +1 @@
+../../TEMPLATE/asr1/db.sh
```
Comment (ftshijt) [inline egs2/lrs2/lipreading1/db.sh:1]: Seems that you forgot to update db.sh for this corpus?

File: egs2/lrs2/lipreading1/local/feature_extract/video_processing.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
+import skvideo.io
+import skimage.transform
```
Comment (ftshijt) [inline egs2/lrs2/lipreading1/local/feature_extract/video_processing.py:2]: Do we already have this package as the requirement of espnet? If not, maybe do the check somewhere before (e.g., data.sh or path.sh)

**Review by ftshijt — COMMENTED** [review]
Looks great! Just please remember to update the corresponding corpus details in egs2/readme.

File: egs2/lrs2/lipreading1/local/feature_extract/video_processing.py (hunk 3 above)
Comment (LiChenda) [reply egs2/lrs2/lipreading1/local/feature_extract/video_processing.py:2]: No, we don't have this. But I add a check in `data.sh` to remind users to install it manually. Do you think it is better to ad d it to the reqirements or we just remind user install it by themself?

File: egs2/lrs2/lipreading1/db.sh (hunk 2 above)
Comment (LiChenda) [reply egs2/lrs2/lipreading1/db.sh:1]: Thanks for you reminding! It's Done, as well as the egs2/readme.

File: egs2/lrs2/lipreading1/local/feature_extract/video_processing.py (hunk 3 above)
Comment (ftshijt) [reply egs2/lrs2/lipreading1/local/feature_extract/video_processing.py:2]: Oh, sorry, I did not see the check. I think the current version is already in a good shape. Thanks

File: egs2/TEMPLATE/asr1/asr.sh [hunk 4]
```diff
@@ -510,6 +510,11 @@ if ! "${skip_data_prep}"; then
                 <data/"${dset}"/cmvn.scp awk ' { print($1,"<DUMMY>") }' > data/"${dset}"/wav.scp
                 utils/copy_data_dir.sh --validate_opts --non-print data/"${dset}" "${data_feats}${_suf}/${dset}"
 
+                # Derive the the frame length and feature dimension
+                _nj=$(min "${nj}" "$(<"${data_feats}${_suf}/${dset}/utt2spk" wc -l)")
+                scripts/feats/feat_to_shape.sh --nj "${_nj}" --cmd "${train_cmd}" \
+                    "${data_feats}${_suf}/${dset}/feats.scp" "${data_feats}${_suf}/${dset}/feats_shape"
```
Comment (sw005320) [inline egs2/TEMPLATE/asr1/asr.sh:516]: Where is `feats_shape` used?
Is it needed?

Comment (LiChenda) [reply egs2/TEMPLATE/asr1/asr.sh:516]: Yes, it is needed by [here](https://github.com/espnet/espnet/blob/06335e6a427e174d1db5e68c97fba944451426f5/egs2/TEMPLATE/asr1/asr.sh#L571) .
In that `else` branch, except `raw` , the other types of freature require `feats_shape`.

**Comment by sw005320** [comment]
Thanks, @LiChenda!
If you get a result, please make an additional PR!

**Pull request merged**